### PR TITLE
feat(agent): CodeAct with function calling

### PR DESCRIPTION
**Pull request by xingyaoww**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

~This PR is a (very) in-complete draft. Several problems need to be addressed, including:~
* ~Most tricky one: How should we extend our *message* system to support tool call. Should we dump the Message abstraction from OpenHands and use abstraction from LiteLLM directly? So we can go from Action/Observation -> LiteLLM Message, instead of need to go through another layer of translating (Action/Observation -> OpenHands Message -> LiteLLM Message)~

Build on top of https://github.com/All-Hands-AI/OpenHands/pull/4528

EDIT: Well - we got it working with 38% SWE-Bench Lite score: https://x.com/gneubig/status/1849874034810618180

---
**Link of any specific issues this addresses**

### CODE REVIEW

File: openhands/agenthub/codeact_agent/function_calling.py
Comment (enyst) [inline openhands/agenthub/codeact_agent/function_calling.py:243]: Oh, do we need this one at all? The MessageAction is default anyway, for CodeAct, when the LLM isn't trying to _do something_. Getting a regular response like some text just works, doesn't it?

Comment (xingyaoww) [reply openhands/agenthub/codeact_agent/function_calling.py:243]: Since we are using function calling anyway - i figured maybe it will be easier if we just fix the action space for the agent AND force LLM to choose between one of them. But i guess what you said also make sense: maybe the model is trained better when responding to assistant message 🤔 maybe I'll revert

Comment (xingyaoww) [reply openhands/agenthub/codeact_agent/function_calling.py]: yeah reverted - maybe this is closer to the training distribution. Thanks for pointing out!

File: openhands/agenthub/codeact_agent/codeact_agent.py
Comment (chnsh) [inline openhands/agenthub/codeact_agent/codeact_agent.py:156]: minor nit: Could be helpful to name this appropriately to improve readability

File: openhands/agenthub/codeact_agent/function_calling.py
Comment (enyst) [inline openhands/agenthub/codeact_agent/function_calling.py:328]:
```suggestion
```

**Comment by enyst** [comment]
Just a quick thought re: Message. If you feel it may be simpler to use litellm's Message, couldn't we get both by subclassing it? We could then just add our prompt caching and stuff, and otherwise let it fall back to litellm's attributes.

**Comment by xingyaoww** [comment]
> {"type":"error","error":{"type":"invalid_request_error","message":"messages.2.content.0.tool_result: cache_control is not supported within tool result content"}}

anthropic prompt caching doesn't support tool calling... such a bummer :(

File: pyproject.toml
Comment (xingyaoww) [inline pyproject.toml:17]: We need to stick with this until litellm merges my PR: https://github.com/BerriAI/litellm/pull/6425 :(

File: openhands/runtime/plugins/agent_skills/file_editor/base.py
Comment (neubig) [inline openhands/runtime/plugins/agent_skills/file_editor/base.py:18]:
```suggestion
```
Maybe delete?

File: openhands/runtime/plugins/agent_skills/file_editor/run.py
Comment (neubig) [inline openhands/runtime/plugins/agent_skills/file_editor/run.py:6]: Just checking, it's intentional that "<response clipped>" is opened but not closed, right?

File: pyproject.toml
Comment (neubig) [inline pyproject.toml:17]: It'd be nice if we didn't have to do this, but maybe we can do it for now and open an issue to remember to revert and share it with the litellm folks?

**Review by neubig — APPROVED** [review]
Looks good, I approve but just have some suggestions.

File: evaluation/integration_tests/tests/t01_fix_simple_typo.py
Comment (rbren) [inline evaluation/integration_tests/tests/t01_fix_simple_typo.py:11]: this seems like an anti-pattern--end users are not going to specify absolute paths...

File: openhands/runtime/plugins/agent_skills/file_editor/base.py
Comment (xingyaoww) [reply openhands/runtime/plugins/agent_skills/file_editor/base.py]: deleted in some of the recent commit!

File: openhands/runtime/plugins/agent_skills/file_editor/run.py
Comment (xingyaoww) [reply openhands/runtime/plugins/agent_skills/file_editor/run.py:6]: I copied this as is from anthropic's quick start repo - so probably yes

File: evaluation/integration_tests/tests/t01_fix_simple_typo.py
Comment (xingyaoww) [reply evaluation/integration_tests/tests/t01_fix_simple_typo.py]: just fixed in the most recent commit

File: openhands/agenthub/codeact_agent/function_calling.py
Comment (enyst) [inline openhands/agenthub/codeact_agent/function_calling.py:28]: Good kind of change, Anthropic does use xml tags so much, even like this ^^.

Comment (xingyaoww) [reply openhands/agenthub/codeact_agent/function_calling.py:28]: haha - this is heavily inspired by their quickstart: https://github.com/anthropics/anthropic-quickstarts/blob/8f734fd08c425c6ec91ddd613af04ff87d70c5a0/computer-use-demo/computer_use_demo/loop.py#L56-L70

**Comment by xingyaoww** [comment]
opps, we need to get https://github.com/All-Hands-AI/OpenHands/pull/4573 in before this

**Comment by xingyaoww** [comment]
I will switch this to draft to prevent accidental merge - I'm doing some model compatibility test to make sure this can work across major providers we recommend (Anthropic/OpenAI/Deepseek).

Also we need to get the following merge first (issues i ran into during eval):
- Agent controller won't stop when Fatal Error occurs: https://github.com/All-Hands-AI/OpenHands/pull/4573
- Better runtime image caching (otherwise tweaking dependencies here is a huge pain - you need to rebuild 700+ images from scratch everytime you change it): https://github.com/All-Hands-AI/OpenHands/pull/4574

File: openhands/agenthub/codeact_agent/function_calling.py
Comment (xingyaoww) [reply openhands/agenthub/codeact_agent/function_calling.py:272]: Revert back after 4 days -- for weaker model like Deepseek, if you don't force it to generate tool call via `tool_choice='required'` - it will be hard to make it generate tool calls 😓 

So i guess the easy way would be make ALL its action space as tool call and force a tool call for *each* action.

**Comment by xingyaoww** [comment]
This is ready for re-review! I implemented something that would allow CodeAct to return multiple actions at once so it can also support models from other providers (e.g., OpenAI models).

I ran SWE-Bench Lite and got these numbers:
- claude-3-5-sonnet-20241022:  # resolved: 125 / 300. (41.67%)
- gpt-4o-2024-08-26: # resolved: 56 / 300. (18.67%)
- gpt-4o-2024-05-13: # resolved: 57 / 300. (19.00%)
- gpt-4o-mini-2024-07-18: # resolved: 17 / 300. (5.67%)

Currently running the Verified set.

File: openhands/agenthub/codeact_agent/codeact_agent.py
Comment (rbren) [inline openhands/agenthub/codeact_agent/codeact_agent.py:106]: all the `if/else` function_calling logic here is a little ugly. I wonder if we just need a new CodeActFunctionAgent which inherits from CodeAct?

Not for this PR necessarily, but it seems like the logic diverges quite a bit

Comment (xingyaoww) [reply openhands/agenthub/codeact_agent/codeact_agent.py:106]: Totally agree here! I think that makes sense - I can put up a follow-up PR to clean this up a bit.

File: evaluation/swe_bench/run_infer.py
Comment (enyst) [inline evaluation/swe_bench/run_infer.py:499]: Hah. Sneaky, but probably good to have. 😅

File: openhands/core/message.py
Comment (enyst) [inline openhands/core/message.py:74]: @xingyaoww  I saw the discussion, but it looks like our code completely removed the single-string serializer? It wasn't just for this, it was needed for LLMs or providers that don't support list of strings/images at all, they just need a single string per message (sadly). I think it's the old version of openAI API, and not everyone has adapted to the new one.

That said, this serializer was intended as temporary. If they adapt or LiteLLM does handle it completely, then we will not need to do this. I'll test to see where we are now.

File: pyproject.toml
Comment (enyst) [inline pyproject.toml:17]: This was for the tool caching issue, right? I'll remove it when we got the version with it.

File: evaluation/swe_bench/run_infer.py
Comment (xingyaoww) [reply evaluation/swe_bench/run_infer.py:499]: i've once been told during one of my internship that "disk are cheap" 🤣

File: pyproject.toml
Comment (xingyaoww) [reply pyproject.toml:17]: yep - i'm just waiting for litellm to cut a new release with the fixed cversion

File: openhands/agenthub/codeact_agent/function_calling.py
Comment (li-boxuan) [inline openhands/agenthub/codeact_agent/function_calling.py:365]: @xingyaoww This code puzzles me a bit. Is it by design?

Comment (xingyaoww) [reply openhands/agenthub/codeact_agent/function_calling.py:365]: yeah, basically, it is possible that an LLM returns multiple tool calls from *one response*.

This basically make treat that respose.content as "thought" and tag it with the first tool call the LLM made

Comment (li-boxuan) [reply openhands/agenthub/codeact_agent/function_calling.py:365]: Gotcha, thanks!

**Pull request merged**
